include/deal.II/base/thread_local_storage.h

```diff
@@ -193,40 +179,59 @@ namespace Threads
   // ----------------- inline and template functions --------------------------
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
-    : data(t)
+  inline ThreadLocalStorage<T>::ThreadLocalStorage()
   {}
 
 
   template <typename T>
-  inline ThreadLocalStorage<T>::ThreadLocalStorage(
-    const ThreadLocalStorage<T> &t)
-    : data(t)
-  {}
+  inline ThreadLocalStorage<T>::ThreadLocalStorage(const T &t)
+  {
+    exemplar = std::make_shared<T>(t);
+  }
 
 
   template <typename T>
   inline T &
-  ThreadLocalStorage<T>::get()
+  ThreadLocalStorage<T>::get(bool &exists)
   {
-#    ifdef DEAL_II_WITH_THREADS
-    return data.local();
-#    else
-    return data;
-#    endif
+    const std::thread::id my_id = std::this_thread::get_id();
+
+    auto it = data.find(my_id);
+    if (it != data.end())
+      {
+        exists = true;
+        return it->second;
+      }
+
+    exists = false;
+
+    std::lock_guard lock(insertion_mutex);
+
+    // FIXME: Port to C++11
+
+    if constexpr (std::is_trivially_copyable<T>::value)
+      if (exemplar)
+        {
+          it = data.insert(std::make_pair(my_id, *exemplar)).first;
+          return it->second;
+        }
+
+    if constexpr (std::is_default_constructible<T>::value)
+      {
+        it = data.insert(std::make_pair(my_id, T())).first;
+        return it->second;
```

Comment:
Question: Does `std::map` guarantee that references to values remain valid? In other words, that `std::map` never moves objects around? That's because you're returning a reference to the map's value here and above, so it better stay at that memory location.

A safer design would be to make the map's value type a `std::unique_ptr` to a `T` and return the deferenced pointer. We know that that remains at that memory location.

Reply:
The STL guarantees that `insert` and `emplace` do not invalidate references and iterators. I would prefer to use a `std::map<..., T>` over `std::map<..., std::unique_ptr<T>>` because the latter doesn't have a copy constructor (that we need, because our `ThreadLocalStorage` offers a copy constructor).

Reply:
Well, but the requirement is that the *iterator* remains stable, not that what you get by derefencing it remains stable. The iterator could contain a pointer to the map and an integer for an array element in which the element is stored, which continues to work if the map decides to re-allocate its array somewhere else. But the dereferenced address may not work any more after that.

Copy operator: We might simply have to do the copying by hand :-)

Reply:
The requirement is precisely that if you hold a reference to a stored object that this reference remains stable.

Reply:
(in particular [res.on.data.races] and [container.requirements.dataraces])